### PR TITLE
Twitter api

### PR DESCRIPTION
**Pull request by jennyadamsnz**
A few things that should/could be done that I'm not 100% sure how to do and would appreciate suggestions on:
- storing keys and tokens in a local.py? Is this needed?
- making the url in the tweet a clickable link - nice to have
- then auto-shortening the url
- setting up a character limit so that if the character limit is over 140 that the "Welcome" part of the string is dropped to give extra characters - probably good to have in case of a long website nice which will mean a long url...

### CODE REVIEW

**Review by ryami333 — COMMENTED** [review]
Because I haven't really worked on a Python/Django project before, I'm not certain - but my gut says that these creds shouldn't be hardcoded in here, they should be environment variables or something. Otherwise, anyone with access to this repo will be able to make tweets on behalf of that account, and they're now forever immortalised in this repo's history 😭 You may need to invalidate these keys, generate new ones, and find a way to use them without committing them.

**Comment by jennyadamsnz** [comment]
I've run a test on my local to the new MWW Twitter profile (currently private) which is here and shows a test run of the current code: https://twitter.com/MadeWithWagtail

**Comment by loicteixeira** [comment]
As agreed last week I'll have a look at this PR later this week, however as @ryami333 pointed, this repo is public, therefore anybody can read the credentials. 

Things to do now:
- Remove the `twitter_credentials.py` file.
- Add the required keys to `local.py.example` so people know that those keys are expected.
- Rebase the branch to remove any trace of the `twitter_credentials.py` file. It should appear that this file never existed.
- Force push the rebase branch to the remote
- Log into Twitter and make the current credentials invalid.

On deployment, we will need to make sure to update `local.py` with the required keys.

**Comment by jennyadamsnz** [comment]
Cool, thanks everyone. Will do today. Fortunately I made the twitter ac private ;) whew!

File: .gitignore [hunk 1]
```diff
@@ -32,7 +32,7 @@ media/*
 .vscode
 .coverage
 .bundle
-
+credentials.py
```
Comment (loicteixeira) [inline .gitignore]: This shouldn't be necessary anymore.

File: madewithwagtail/settings/local.py.example [hunk 2]
```diff
@@ -42,3 +42,11 @@ WAGTAILSEARCH_BACKENDS = {
...
     },
 }
+
+#Twitter credentials
+
+# Create variables for each key, secret, token
+CONSUMER_KEY = ''
+CONSUMER_SECRET = ''
+ACCESS_TOKEN = ''
+ACCESS_TOKEN_SECRET = ''
```
Comment (loicteixeira) [inline madewithwagtail/settings/local.py.example]: Please prefix the 4 constants with `TWITTER_` to keep things tidy.

File: core/signals.py [hunk 3]
```diff
@@ -112,3 +115,53 @@ def send_to_slack(sender, **kwargs):
...
+    if is_first_publish:
+
+        auth = tweepy.OAuthHandler(settings.CONSUMER_KEY, settings.CONSUMER_SECRET)
+        auth.set_access_token(settings.ACCESS_TOKEN, settings.ACCESS_TOKEN_SECRET)
+
+        api = tweepy.API(auth)
+
+        maxTweetLength = 140
+        prefix = "New site on MWW! Welcome '"
+        url = "https://www.sp.co.nz/"
```
Comment (loicteixeira) [inline core/signals.py]: This should be `url = page.url`

File: core/signals.py [hunk 4]
```diff
@@ -112,3 +115,53 @@ def send_to_slack(sender, **kwargs):
...
+            endOfWordIndex = max([truncatedTitle.rfind(' '), truncatedTitle.rfind('.')])
+            truncatedTitle = truncatedTitle[0:endOfWordIndex]
+            title = truncatedTitle + ellipsis
+        else:
+            title = page.title
+
+        tweet = prefix + title + titleSuffix + url + handle
+
+        # full tweet format that includes the status and image
+        api.update_with_media(page.image_desktop.file.name, tweet, file=page.image_desktop.file);
```
Comment (loicteixeira) [inline core/signals.py]: `WagtailSitePage.image_desktop` isn't a required field and might be `None`. I know we normally set one but the code should send a text only tweet if there is no image.

File: core/signals.py [hunk 5]
```diff
@@ -112,3 +115,53 @@ def send_to_slack(sender, **kwargs):
...
+        titleSuffix = "' "
+        ellipsis = '\u2026'
+        handle = ""
+
+        # if the submission contains a twitter handle, add this to the end of the tweet
+        parent_page = page.get_parent().specific
+        if parent_page.twitter_handler:
+            handle = " by " + parent_page.twitter_handler
+
+        urlMaxSize = 23
```
Comment (loicteixeira) [inline core/signals.py]: Where does that number comes from?

File: madewithwagtail/settings/base.py [hunk 6]
```diff
@@ -209,3 +209,5 @@
     'DEFAULT_PERMISSION_CLASSES': ('rest_framework.permissions.IsAuthenticatedOrReadOnly',),
     'PAGINATE_BY': None,
 }
+
+from .twitter_credentials import CONSUMER_KEY, CONSUMER_SECRET, ACCESS_TOKEN, ACCESS_TOKEN_SECRET
```
Comment (loicteixeira) [inline madewithwagtail/settings/base.py]: This shouldn't be there.

File: core/signals.py (hunk 4 above)
Comment (loicteixeira) [reply core/signals.py]: Also, doesn't an image take some of the character length away?

**Review by loicteixeira — CHANGES_REQUESTED** [review]
Nice work 👌

How do you like Python? 😄 

I haven't had time to run the code so hopefully my comments aren't too out of place.

File: .gitignore (hunk 1 above)
Comment (jennyadamsnz) [reply .gitignore]: done

File: core/signals.py (hunk 3 above)
Comment (jennyadamsnz) [reply core/signals.py]: yeah thanks, leftover from testing

File: core/signals.py (hunk 5 above)
Comment (jennyadamsnz) [reply core/signals.py]: apparently twitter truncates urls to 23 characters automatically but this is set as a back up :) https://support.twitter.com/articles/78124

File: madewithwagtail/settings/base.py (hunk 6 above)
Comment (jennyadamsnz) [reply madewithwagtail/settings/base.py]: thx

File: madewithwagtail/settings/local.py.example (hunk 2 above)
Comment (jennyadamsnz) [reply madewithwagtail/settings/local.py.example]: ok, will do

File: core/signals.py (hunk 4 above)
Comment (jennyadamsnz) [reply core/signals.py]: Re char length, images don't count as of Sept 2016 so all good there.
Re image requirement, ok i'll check this out if I have time but I finish up tomorrow so no promises ;)

**Comment by jennyadamsnz** [comment]
Thanks @loicteixeira for your reviews and @ryami333 for your help. Unfortunately I've now run out of time to do the final unit testing needed to finish this little project off. @thibaudcolas or @haydnnewport is that something I can leave with one of you please? I would love to see it finally finished!

**Comment by loicteixeira** [comment]
Nice work @jennyadamsnz.

It's sad to see you leave but this project is public so you're more than welcome to continue the work on it if you want to complete it yourself (although you might need to either fork and create a new PR or ask infrastructure to give you write rights to this repo).

Otherwise don't worry, we will wrap it up and merge it :)

**Pull request closed without merging**